### PR TITLE
Implement Modals

### PR DESCRIPTION
**Pull request by CottageDwellingCat**
### Without the interaction service
```cs
var mb = new ModalBuilder()
    .WithTitle("Fav Food")
    .WithCustomId("food_menu")
    .AddTextInput("What??", "food_name", placeholder:"Pizza")
    .AddTextInput("Why??", "food_reason", TextInputStyle.Paragraph, "Kus it's so tasty");

await command.RespondWithModalAsync(mb.Build());
```
```cs
client.ModalSubmitted += async modal => 
{
    List<SocketMessageComponentData> components =
        modal.Data.Components.ToList();
    string food = components
        .Where(x => x.CustomId == "food_name").First().Value;
    string reason = components
        .Where(x => x.CustomId == "food_reason").First().Value;

    string message = "hey @everyone; I just learned " + 
        $"{modal.User.Mention}'s favorite food is " +
        $"{food} because {reason}.";

    AllowedMentions mentions = new AllowedMentions();
    mentions.AllowedTypes = AllowedMentionTypes.Users;

    await modal.RespondAsync(message, allowedMentions:mentions);
}
```

### With the interaction service (sample from cenngo)

```cs
[SlashCommand("food", "Tell us about your favorite food.")]
public async Task Command()
    => await Context.Interaction.RespondWithModalAsync<FoodModal>("food_menu");
        
[ModalInteraction("food_menu")]
public async Task ModalResponce(FoodModal modal)
{
    // Build the message to send.
    string message = "hey @everyone, I just learned " +
        $"{Context.User.Mention}'s favorite food is " +
        $"{modal.Food} because {modal.Reason}.";

    // Specify the AllowedMentions so we don't actually ping everyone.
    AllowedMentions mentions = new();
    mentions.AllowedTypes = AllowedMentionTypes.Users;

    // Respond to the modal.
    await RespondAsync(message, allowedMentions: mentions, ephemeral:true);
}

public class FoodModal : IModal
{
    public string Title => "Fav Food";
    // Strings with the ModalTextInput attribute will automatically 
    [InputLabel("What??")]
    [ModalTextInput("food_name", placeholder:"Pizza", maxLength:20)]
    public string Food { get; set; }
        
    // Additional paremeters can be specified to further customize the input.
    [InputLabel("Why??")]
    [ModalTextInput("food_reason", TextInputStyle.Paragraph, "Kuz it's tasty", maxLength:500)]
    public string Reason { get; set; } 
}
```

## How It Looks
image3
image4

### CODE REVIEW

File: docs/guides/int_basics/message-components/text-input.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+Text input components are a type of MessageComponents that can only be 
+used in modals. Texts inputs can be longer (the `Paragraph`) style or 
+shorter (the `Short` style). Text inputs have a variable min and max 
+length.
+
+![A modal with short and paragraph text inputs](images/image7.png)
+
+## Creating text inputs
+Text input components can be built using the `TextInputBuilder`.
+The simplest text input can build with:
```
Comment (lhjt) [inline docs/guides/int_basics/message-components/text-input.md]:
```suggestion
The simplest text input can built with:
```

File: docs/guides/int_basics/message-components/text-input.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+```cs
+var tb = new TextInputBuilder()
+    .WithLabel("My Text")
+    .WithCustomId("text_input");
+```
+
+and would produce a component looking like:
```
Comment (lhjt) [inline docs/guides/int_basics/message-components/text-input.md]:
```suggestion
and would produce a component that looks like:
```

File: docs/guides/int_basics/modals/intro.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+title: Getting Started with Modals
+---
+# Modals
+
+> [!WARNING]
+> Modals are currently not supported on Android.
+
+## Getting started with modals
+This guide will show you how to use modals and give a few examples of
+valid usecases. If you questions not covered by this guide ask in the
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
valid use cases. If your question is not covered by this guide ask in the
```

File: docs/guides/int_basics/modals/intro.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+### What is a modal?
+Modals are forms bots can send when responding to interactions. Modals 
+are sent to Discord as an array of message components and converted 
+into the form layout by user's clients. Modals are required to have a
+custom id, title, and at least one component. 
+
+![Screenshot of a modal](images/image2.png)
+
+When users submit modals your client fires the ModalSubmitted event. 
+You can get the modals components from the `Data.Components` property
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
You can get the components of the modal from the `Data.Components` property
```

File: docs/guides/int_basics/modals/intro.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+### What is a modal?
+Modals are forms bots can send when responding to interactions. Modals 
+are sent to Discord as an array of message components and converted 
+into the form layout by user's clients. Modals are required to have a
+custom id, title, and at least one component. 
+
+![Screenshot of a modal](images/image2.png)
+
+When users submit modals your client fires the ModalSubmitted event.
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
When users submit modals, your client fires the ModalSubmitted event. 
```

File: docs/guides/int_basics/modals/intro.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+command that will respond with the modal.
+```cs
+[SlashCommand("food", "Tell us about your favorite food!")]
+public async Task FoodPreference()
+{
+    // send a modal
+}
+```
+
+Now that we have our command setup we need to build a modal.
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
Now that we have our command set up, we need to build a modal.
```

File: docs/guides/int_basics/modals/intro.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    .WithCustomId("food_menu")
+    .AddTextInput("What??", "food_name", placeholder:"Pizza")
+    .AddTextInput("Why??", "food_reason", TextInputStyle.Paragraph, 
+        "Kus it's so tasty");
+    
+    await Context.Interaction.RespondWithModalAsync(mb.Build());
+}
+```
+
+When we run the command our modal should pop up:
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
When we run the command, our modal should pop up:
```

File: docs/guides/int_basics/modals/intro.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+![screenshot of the above modal](images/image3.png) 
+
+### Respond to modals
+
+> [!WARNING]
+> Modals can not be sent when respoding to a modal.
+
+Once a user has submitted our modal we need to let everyone know what
+their favorite food is. We can start by hooking a task to the clients
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
Once a user has submitted the modal, we need to let everyone know what
their favorite food is. We can start by hooking a task to the client's
```

File: docs/guides/int_basics/modals/intro.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    // Respond to the modal.
+    await modal.RespondAsync(message, allowedMentions:mentions);
+}
+```
+
+Now responding to the modal should inform everyone of our tasty 
+choices.
+
+![Responce of the modal submitted event](images/image4.png)
```
Comment (lhjt) [inline docs/guides/int_basics/modals/intro.md]:
```suggestion
![Response of the modal submitted event](images/image4.png)
```

File: docs/guides/int_framework/intro.md [hunk 10]
```diff
@@ -198,6 +198,18 @@ Autocomplete commands must be parameterless methods. A valid Autocomplete comman
 
 Alternatively, you can use the [AutocompleteHandlers] to simplify this workflow.
 
+## Modals
+
+Modal commands fist paremeter must be an implimentation of `IModal`.
```
Comment (lhjt) [inline docs/guides/int_framework/intro.md]:
```suggestion
Modal commands first parameter must be an implementation of `IModal`.
```

File: docs/guides/int_framework/intro.md [hunk 11]
```diff
@@ -198,6 +198,18 @@ Autocomplete commands must be parameterless methods. A valid Autocomplete comman
...
+An Modal implimentation would look like:
```
Comment (lhjt) [inline docs/guides/int_framework/intro.md]:
```suggestion
A Modal implementation would look like this:
```

File: docs/guides/int_framework/samples/intro/modal.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        AllowedMentions mentions = new();
+        mentions.AllowedTypes = AllowedMentionTypes.Users;
+
+        // Respond to the modal.
+        await RespondAsync(message, allowedMentions: mentions);
+    }
+    
+    public class FoodModal : IModal
+    {
+        // Title and CrustomId are both required.
```
Comment (lhjt) [inline docs/guides/int_framework/samples/intro/modal.cs]:
```suggestion
        // Title and CustomId are both required.
```

File: src/Discord.Net.Core/Entities/Interactions/MessageComponents/IComponentInteractionData.cs [hunk 13]
```diff
@@ -21,5 +21,10 @@ public interface IComponentInteractionData : IDiscordInteractionData
         ///     Gets the value(s) of a <see cref="SelectMenuComponent"/> interaction response.
         /// </summary>
         IReadOnlyCollection<string> Values { get; }
+
+        /// <summary>
+        ///     Gets the value of a <see cref="TextInputComponent"/> interaction responce.
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/MessageComponents/IComponentInteractionData.cs]:
```suggestion
        ///     Gets the value of a <see cref="TextInputComponent"/> interaction response.
```

File: src/Discord.Net.Core/Entities/Interactions/MessageComponents/TextInputComponent.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    public class TextInputComponent : IMessageComponent
+    {
+        /// <inheritdoc/>
+        public ComponentType Type => ComponentType.TextInput;
+
+        /// <inheritdoc/>
+        public string CustomId { get; }
+
+        /// <summary>
+        ///     Gets the label of the component, this is the text shown above it.
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/MessageComponents/TextInputComponent.cs]:
```suggestion
        ///     Gets the label of the component; this is the text shown above it.
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/IModalInteraction.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+namespace Discord
+{
+    /// <summary>
+    ///     Represents an interaction type for Modals.
+    /// </summary>
+    public interface IModalInteraction : IDiscordInteraction
+    {
+        /// <summary>
+        ///     Gets the data received with this interaction, contains the button that was clicked.
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/IModalInteraction.cs]:
```suggestion
        ///     Gets the data received with this interaction; contains the clicked button.
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Discord
+{
+    public class ModalBuilder
+    {
+        /// <summary>
+        ///     Gets or sets the components of the current model.
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        ///     Gets or sets the components of the current modal.
```
Potentially?

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+        /// </summary>
+        public ModalComponentBuilder Components { get; set; } = new();
+
+        /// <summary>
+        ///     Gets or sets the title of the current modal.
+        /// </summary>
+        public string Title { get; set; }
+
+        /// <summary>
+        ///     Gets or sets the custom id of the current modal..
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        ///     Gets or sets the custom id of the current modal.
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+        private string _customId;
+
+        public ModalBuilder() { }
+
+        /// <summary>
+        ///     Creates a new instance of a <see cref="ModalBuilder"/>
+        /// </summary>
+        /// <param name="title">The modals title.</param>
+        /// <param name="customId">The modals customId.</param>
+        /// <param name="components">The modals components.</param>
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        /// <param name="title">The modal's title.</param>
        /// <param name="customId">The modal's customId.</param>
        /// <param name="components">The modal's components.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+        /// <exception cref="ArgumentException">Only TextInputComponents are allowed.</exception>
+        public ModalBuilder(string title, string customId, ModalComponentBuilder components = null)
+        {
+            Title = title;
+            CustomId = customId;
+            Components = components ?? new();
+        }
+
+        /// <summary>
+        ///     Sets the title title of the current modal.
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        ///     Sets the title of the current modal.
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+
+        /// <summary>
+        ///     Adds a <see cref="TextInputBuilder"/> to the current builder.
+        /// </summary>
+        /// <param name="customId">The inputs custom id.</param>
+        /// <param name="label">The inputs label.</param>
+        /// <param name="placeholder">The inputs placeholder text.</param>
+        /// <param name="minLength">The inputs minimum length.</param>
+        /// <param name="maxLength">The inputs maximum length.</param>
+        /// <param name="style">The inputs style.</param>
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        /// <param name="customId">The input's custom id.</param>
        /// <param name="label">The input's label.</param>
        /// <param name="placeholder">The input's placeholder text.</param>
        /// <param name="minLength">The input's minimum length.</param>
        /// <param name="maxLength">The input's maximum length.</param>
        /// <param name="style">The input's style.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+        /// <summary>
+        ///     Adds a <see cref="TextInputBuilder"/> to the <see cref="ComponentBuilder"/> at the specific row.
+        ///     If the row cannot accept the component then it will add it to a row that can.
+        /// </summary>
+        /// <param name="customId">The inputs custom id.</param>
+        /// <param name="label">The inputs label.</param>
+        /// <param name="placeholder">The inputs placeholder text.</param>
+        /// <param name="minLength">The inputs minimum length.</param>
+        /// <param name="maxLength">The inputs maximum length.</param>
+        /// <param name="style">The inputs style.</param>
```
Comment (lhjt) [inline src/Discord.Net.Core/Entities/Interactions/Modals/ModalBuilder.cs]:
```suggestion
        /// <param name="customId">The input's custom id.</param>
        /// <param name="label">The input's label.</param>
        /// <param name="placeholder">The input's placeholder text.</param>
        /// <param name="minLength">The input's minimum length.</param>
        /// <param name="maxLength">The input's maximum length.</param>
        /// <param name="style">The input's style.</param>
```

File: src/Discord.Net.Interactions/Attributes/Commands/ModalInteractionAttribute.cs [hunk 22]
```diff
@@ -0,0 +1,44 @@
+using System;
+
+namespace Discord.Interactions
+{
+    /// <summary>
+    ///     Create a Modal interaction handler, CustomId represents
```
Comment (lhjt) [inline src/Discord.Net.Interactions/Attributes/Commands/ModalInteractionAttribute.cs]:
```suggestion
    ///     Create a Modal interaction handler. CustomId represents
```

File: src/Discord.Net.Interactions/Attributes/Modals/RequiredInputAttribute.cs [hunk 23]
```diff
@@ -0,0 +1,25 @@
+using System;
+
+namespace Discord.Interactions
+{
+    /// <summary>
+    ///     Sets the input as required or optinal.
```
Comment (lhjt) [inline src/Discord.Net.Interactions/Attributes/Modals/RequiredInputAttribute.cs]:
```suggestion
    ///     Sets the input as required or optional.
```

File: src/Discord.Net.Interactions/InteractionServiceConfig.cs [hunk 24]
```diff
@@ -56,6 +59,11 @@ public class InteractionServiceConfig
         ///     Gets or sets delegate to be used by the <see cref="InteractionService"/> when responding to a Rest based interaction.
         /// </summary>
         public RestResponseCallback RestResponseCallback { get; set; } = (ctx, str) => Task.CompletedTask;
+
+        /// <summary>
+        ///     Gets or sets whether an exception should be thrown when a component does not belong to a field in a modals class.
```
Comment (lhjt) [inline src/Discord.Net.Interactions/InteractionServiceConfig.cs]:
```suggestion
        ///     Gets or sets whether an exception should be thrown when a component does not belong to a field in a modal's class.
```
Potentially?

File: src/Discord.Net.WebSocket/Entities/Interaction/MessageComponents/SocketMessageComponentData.cs [hunk 25]
```diff
@@ -23,11 +23,31 @@ public class SocketMessageComponentData : IComponentInteractionData
         /// </summary>
         public IReadOnlyCollection<string> Values { get; }
 
+        /// <summary>
+        ///     Gets the value of a <see cref="TextInputComponent"/> interaction responce.
```
Comment (lhjt) [inline src/Discord.Net.WebSocket/Entities/Interaction/MessageComponents/SocketMessageComponentData.cs]:
```suggestion
        ///     Gets the value of a <see cref="TextInputComponent"/> interaction response.
```

File: src/Discord.Net.WebSocket/Entities/Interaction/Modals/SocketModalData.cs [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+    /// </summary>
+    public class SocketModalData : IDiscordInteractionData, IModalInteractionData
+    {
+        /// <summary>
+        ///     Gets the <see cref="Modal"/>'s Custom Id.
+        /// </summary>
+        public string CustomId { get; }
+
+        /// <summary>
+        ///     Gets the <see cref="Modal"/>s components submitted by the user.
```
Comment (lhjt) [inline src/Discord.Net.WebSocket/Entities/Interaction/Modals/SocketModalData.cs]:
```suggestion
        ///     Gets the <see cref="Modal"/>'s components submitted by the user.
```

**Review by lhjt — CHANGES_REQUESTED** [review]
The guides look pretty good 👍

**Review by lhjt — APPROVED** [review]

File: src/Discord.Net.Interactions/InteractionServiceConfig.cs [hunk 27]
```diff
@@ -56,6 +59,11 @@ public class InteractionServiceConfig
         ///     Gets or sets delegate to be used by the <see cref="InteractionService"/> when responding to a Rest based interaction.
         /// </summary>
         public RestResponseCallback RestResponseCallback { get; set; } = (ctx, str) => Task.CompletedTask;
+
+        /// <summary>
+        ///     Gets or sets whether an exception should be thrown when a component does not belong to a field in a modal's class.
+        /// </summary>
+        public bool ThrowOnUnknownModalComponent { get; set; } = false;
```
Comment (Cenngo) [inline src/Discord.Net.Interactions/InteractionServiceConfig.cs]: oops, seems like i forgot to implement this. ill open a pr for it on your branch.

File: src/Discord.Net.Interactions/Info/ModalInfo.cs [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            Title = builder.Title;
+            Type = builder.Type;
+            Components = builder.Components.Select(x => x switch
+            {
+                Builders.TextInputComponentBuilder textComponent => textComponent.Build(this),
+                _ => throw new InvalidOperationException($"{x.GetType().FullName} isn't a supported modal input component builder type.")
+            }).ToImmutableArray();
+
+            TextComponents = Components.Where(x => x is TextInputComponentInfo).Cast<TextInputComponentInfo>().ToImmutableArray();
```
Comment (Cenngo) [inline src/Discord.Net.Interactions/Info/ModalInfo.cs]: @CottageDwellingCat when you get the chance could you replace this mess with .OfType().ToImmutableArray(), my bad i forgot that method even existed. sry im out of town and i dont have access to a pc

File: src/Discord.Net.Interactions/Builders/Modals/Inputs/IInputComponentBuilder.cs [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+{
+    /// <summary>
+    ///     Represent a builder for creating <see cref="InputComponentInfo"/>.
+    /// </summary>
+    public interface IInputComponentBuilder
+    {
+        /// <summary>
+        ///     Gets the parent modal of this input component.
+        /// </summary>
+        public ModalBuilder Modal { get; }
```
Comment (quinchs) [inline src/Discord.Net.Interactions/Builders/Modals/Inputs/IInputComponentBuilder.cs]: Should interface properties be public?

**Review by quinchs — CHANGES_REQUESTED** [review]

**Comment by SylveonDeko** [comment]
pog

File: src/Discord.Net.Interactions/Builders/Modals/Inputs/IInputComponentBuilder.cs (hunk 29 above)
Comment (CottageDwellingCat) [reply src/Discord.Net.Interactions/Builders/Modals/Inputs/IInputComponentBuilder.cs]: I changed them to be private and consistent with [commands](https://github.com/Discord-Net-Labs/Discord.Net-Labs/blob/a94f9851c958b35955b692ece52385ccc00aa8ed/src/Discord.Net.Interactions/Builders/Commands/ICommandBuilder.cs).

**Review by quinchs — APPROVED** [review]

**Pull request merged**